extra/snake-game/snake-game.factor

```diff
@@ -0,0 +1,395 @@
+    { next-turn-dir initial: f }
+    { score integer initial: 0 }
+    { paused? boolean initial: t }
+    { game-over? boolean initial: f } ;
+
+TUPLE: snake-part
+    dir type ;
+
+: <snake-part> ( dir type -- snake-part )
+    snake-part boa ;
+
+: <snake> ( -- snake )
+    [
+        :left :head <snake-part> ,
+        :left :body <snake-part> ,
+        :left :tail <snake-part> ,
+    ] V{ } make ;
+
+: <snake-game> ( -- snake-game )
+    snake-game new
+    <snake> >>snake
+    { 5 4 } clone >>snake-loc
+    :right >>snake-dir
+    { 1 1 } clone >>food-loc ;
+
+TUPLE: snake-gadget < gadget
+    snake-game timer textures ;
+
+: start-new-game ( snake-gadget -- )
+    <snake-game> >>snake-game drop ;
+
+: <snake-gadget> ( -- snake-gadget )
+    snake-gadget new
+    [ start-new-game ] keep ;
+
+: opposite-dir ( dir -- dir )
+    H{
+        { :left  :right }
+        { :right :left }
+        { :up    :down }
+        { :down  :up }
+    } at ;
+
+: lookup-texture ( key -- texture )
+    game-textures get at ;
+
+: screen-loc ( loc -- loc )
+    [ 20 * ] map ;
+
+: draw-sprite* ( key screen-loc -- )
+    [ lookup-texture draw-texture ] with-translation ;
+
+: draw-sprite ( grid-loc key -- )
+    swap screen-loc draw-sprite* ;
+
+: draw-food ( loc -- )
+    "food" draw-sprite ;
+
+: draw-background ( -- )
+    { 0 0 } "background" draw-sprite ;
+
+: offset ( loc dim -- loc )
+    [ + ] 2map ;
+
+: draw-snake-head ( loc facing-dir -- )
+    dup name>> rest "head-" prepend
+    [
+        [ screen-loc ] dip
+        {
+            { :right [ { -20 -10 } ] }
+            { :down  [ { -10 -20 } ] }
+            { :up    [ { -10  0  } ] }
+            { :left  [ {  0  -10 } ] }
+        } case offset
+    ] dip
+    swap draw-sprite* ;
+
+: draw-snake-body ( loc from-dir to-dir -- )
+    2array [ name>> rest ] map "body" prefix "-" join
+    draw-sprite ;
+
+: draw-snake-tail ( loc facing-dir -- )
+    name>> rest "tail-" prepend draw-sprite ;
+
+: draw-snake-part ( loc from-dir snake-part -- )
+    dup type>> {
+        { :head [ drop opposite-dir draw-snake-head ] }
+        { :tail [ drop draw-snake-tail ] }
+        { :body [ dir>> draw-snake-body ] }
+    } case ;
+
+: ?roll-over ( x max -- x )
+    {
+        { [ 2dup >= ] [ 2drop 0 ] }
+        { [ over neg? ] [ nip 1 - ] }
+        [ drop ]
+    } cond ;
+
+: ?roll-over-x ( x -- x )
+    snake-game-dim first ?roll-over ;
+
+: ?roll-over-y ( y -- y )
+    snake-game-dim second ?roll-over ;
+
+: move ( loc dim -- loc )
+    offset first2
+    [ ?roll-over-x ] [ ?roll-over-y ] bi* 2array ;
+
+: relative-loc ( loc dir -- loc )
+    {
+        { :left  [ { -1  0 } move ] }
+        { :right [ {  1  0 } move ] }
+        { :up    [ {  0 -1 } move ] }
+        { :down  [ {  0  1 } move ] }
+    } case ;
+
+: draw-snake-reduce-step ( loc from-dir snake-part -- {new-loc,new-from-dir} )
+    nip dir>> [ relative-loc ] keep 2array ;
+
+: draw-snake ( snake loc from-dir -- )
+    2array 2dup
```

Comment:
This word might be cleaner if you used ``each`` instead of ``reduce`` and left the ``new-loc`` and ``new-from-dir`` on the stack (so you don't have to ``2array`` and ``first2``).